srml/democracy/src/lib.rs

```diff
@@ -982,6 +982,24 @@ impl<T: Trait> OnFreeBalanceZero<T::AccountId> for Module<T> {
 	}
 }
 
+fn safe_append_value<V, S>(_: S, value: V)
+where
+	V: Encode + Decode + Clone,
+	S: StorageValue<Vec<V>, Query=Vec<V>>,
+{
+	S::append(&[value.clone()]).unwrap_or_else(|_| S::mutate(|st| st.push(value)))
+}
+
+fn safe_append_map<K, V, S>(_: S, key: K, value: V)
+where
+	K: Encode + Decode + Copy,
+	V: Encode + Decode + Clone,
+	S: AppendableStorageMap<K, Vec<V>, Query=Vec<V>>,
+{
+	S::append(key, &[value.clone()])
+			.unwrap_or_else(|_| S::mutate(key, |v| v.push(value)));
```

Comment:
I'm not sure I see the interest of this safe method here; append only fails if value is wrongly encoded.

Also it seems append function returns an error but get function actually panics if value is wrongly encoded. I don't know what we should prefer but this *safe* is actually just calling the other behavior which would panic instead of returning an error.

Reply:
Your point holds, and I am not fully aware of when/how storage actually panics tbh. I've added a new version which is more or less the same but abstracted in the storage layer. But I personally documented it in a way to encourage the users to not use it and `unwrap_or_blah_blah()` themselves. 

But given your note: if `::get` and hence `::mutate()` can _only_ panic in the above code, for the same reason that the original `::append()` failed (corrupted encoding), then this whole `_safe()` is moot. Whenever `::append()` fails, the following `::mutate()` (or `::set()` in the new version) will also fails. Is this true @bkchr ?

Reply:
if the above is correct then it should be called `fn append_or_set(items)` to not give the misunderstanding that it is actually _safer_. It seems to be not.

Reply:
at the same time append already does set if empty.

Reply:
We are currently panicking, but this is not correct. We need to remove this panic. Otherwise you could end up with a corrupt state, when you have something that changed and you forgot to upgrade it in `on_initialize`.
As I said in my comment somewhere here, you should use `Self::put` when `append` fails. 

Hmm, but you probably can not use `put` as you would need to construct the actual type out of the items...

Reply:
Well, I do `::put()` now (already pushed) with the extra where clause of `T: From<&'a[I]>` (e.g. `Vec<T> from &[T]`). Since I think 99% of the use cases of this method is for arrays and vec and they have this `From<&[_]>`  impl out of the box, it should be fine. Otherwise, the use has to implement this somewhere to be able to call `safe_append()`